include/roboptim/core/detail/structured-input.hh

```diff
@@ -0,0 +1,114 @@
+      typedef typename FuncType::jacobian_ref JacBlock;
+      /// \brief input type of the jacobian given by the user
+      typedef typename FuncType::jacobian_ref InputJacBlock;
+
+      /// \brief retrieve a specific block of the jacobian given as an input
+      /// \param jacobian the jacobian to operate upon
+      /// \param blockInd the index of the block to be retrieved
+      JacBlock getJacobianBlock(InputJacBlock jacobian, size_t blockInd) const;
+    };
+
+    /// \brief DenseMatrix specialization of the getJacobianBlock() method provider.
+    template<typename FuncType>
+    struct StructuredInputJacobianInternal<FuncType, roboptim::EigenMatrixSparse> : public BlockProvider
+    {
+      /// \brief return type of the getJacobianBlock() method
+      typedef Eigen::Block<Eigen::SparseMatrix<double>, -1, -1, 0> JacBlock;
+      /// \brief input type of the jacobian given by the user
+      typedef Eigen::SparseMatrix<double>& InputJacBlock;
+
+      /// \brief retrieve a specific block of the jacobian given as an input
+      /// \param jacobian the jacobian to operate upon
+      /// \param blockInd the index of the block to be retrieved
+      JacBlock getJacobianBlock(InputJacBlock jacobian, size_t blockInd) const;
+    };
+
+    /// \brief Provides utility methods to describe the input format of a function
+    ///
+    /// Inheriting from this class will allow a function to define "blocks" of input.
+    /// You can then retrieve a specific block from the whole input, or the specific
+    /// part of the jacobian depending on this block.
+    ///
+    /// To inherit from this class, please specify as the template argument the roboptim function type of the function (Function, DifferentiableFunction, LinearFunction, et cetera)
+    template<typename FuncType>
+    class StructuredInput
+    // This boost::conditional allows us to not define the getJacobianBlock method
+    // if the function is not differentiable.
+      : public boost::conditional<
+      boost::is_base_of<roboptim::GenericDifferentiableFunction<typename FuncType::traits_t>, FuncType>::value,
+      StructuredInputJacobianInternal<FuncType, typename FuncType::traits_t>,
+      BlockProvider>
+    ::type
+    {
+    public:
+      /// \brief Adds a new block of input to the function
+      /// \param size size of the block to add
+      void addBlock(size_t size);
+      /// \brief Returns the number of blocks defined by the function
+      size_t getNumBlocks() const;
+
+      /// \brief return type of the getInputBlock() method
+      typedef Eigen::Ref<const Eigen::VectorXd>::ConstSegmentReturnType ConstSegment;
+
+      /// \brief Reads a specified block of data from an input argument
+      ///
+      /// \param input the input data to be read
+      /// \param blockInd the index of the block to be read
+      ConstSegment getInputBlock(typename FuncType::const_argument_ref input, size_t blockInd) const;
+    };
+
+# include "structured-input.hxx"
```

Comment:
`# include <roboptim/core/detail/structured-input.hxx>` to comply with the rest of the code.